src/chips/safe_accumulator.rs

```diff
@@ -0,0 +1,211 @@
+    pub instance: Column<Instance>,
+    pub is_zero: IsZeroConfig,
+    pub selector: [Selector; 2],
+}
+
+#[derive(Debug, Clone)]
+pub struct SafeACcumulatorChip<const MAX_BITS: u8, const ACC_COLS: usize> {
+    config: SafeAccumulatorConfig<MAX_BITS, ACC_COLS>,
+}
+
+impl<const MAX_BITS: u8, const ACC_COLS: usize> SafeACcumulatorChip<MAX_BITS, ACC_COLS> {
+    pub fn construct(config: SafeAccumulatorConfig<MAX_BITS, ACC_COLS>) -> Self {
+        Self { config }
+    }
+
+    pub fn configure(
+        meta: &mut ConstraintSystem<Fp>,
+        update_value: Column<Advice>,
+        left_most_inv: Column<Advice>,
+        add_carries: [Column<Advice>; ACC_COLS],
+        accumulate: [Column<Advice>; ACC_COLS],
+        selector: [Selector; 2],
+        instance: Column<Instance>,
+    ) -> SafeAccumulatorConfig<MAX_BITS, ACC_COLS> {
+        let add_carry_selector = selector[0];
+        let overflow_check_selector = selector[1];
+
+        let is_zero = IsZeroChip::configure(
+            meta,
+            |meta| meta.query_selector(overflow_check_selector),
+            |meta| meta.query_advice(accumulate[0], Rotation::cur()),
+            left_most_inv,
+        );
+
+        // Enable equality on the advice and instance column to enable permutation check
+        meta.enable_equality(instance);
+        meta.enable_equality(update_value);
+        meta.enable_equality(left_most_inv);
+        accumulate.map(|col| meta.enable_equality(col));
+        add_carries.map(|col| meta.enable_equality(col));
+
+        meta.create_gate("accumulation constraint", |meta| {
+            let s_add = meta.query_selector(add_carry_selector);
+            let s_over = meta.query_selector(overflow_check_selector);
+
+            // value to be added
+            let value = meta.query_advice(update_value, Rotation::cur());
+
+            let previous_acc = (0..ACC_COLS)
+                .map(|i| meta.query_advice(accumulate[i], Rotation::prev()))
+                .collect::<Vec<Expression<Fp>>>();
+            let carries_acc = (0..ACC_COLS)
+                .map(|i| meta.query_advice(add_carries[i], Rotation::cur()))
+                .collect::<Vec<Expression<Fp>>>();
+            let updated_acc = (0..ACC_COLS)
+                .map(|i| meta.query_advice(accumulate[i], Rotation::cur()))
+                .collect::<Vec<Expression<Fp>>>();
+
+            let shift_next_chunk = Expression::Constant(Fp::from(1 << MAX_BITS));
+
+            // Add the value with right most accumulation column
+            let check_add_value_exprs = vec![
+                s_add.clone()
+                    * ((value + previous_acc[ACC_COLS - 1].clone())
+                        - ((carries_acc[ACC_COLS - 1].clone() * shift_next_chunk.clone())
+                            + updated_acc[ACC_COLS - 1].clone())),
+            ];
+            // Check with other accumulation columns with carries
+            let check_accumulates_with_carries_expr = (1..ACC_COLS - 1)
```

Comment:
Can you better explain this constraint?

Reply:
In same circuit configuration above, starting with left most of accumulate columns, `accumulate_2`.

| -             | new_value | left_most_inv | add_carries_2 | add_carries_1 | add_carries_0 | >accumulate_2< | accumulate_1 | accumulate_0 | add_selector |
| --            | --        | --            | --            | --            | --            | --           | --           | -            | -            |
| previous_acc  |           |               |               |               |               | 0 ←          | 0xe          | 0xd          | 0            | 
| updated_acc   | 0x5       | 0             | → 0           | 0 ←           | 1             | → 0          | 0xf          | 1            | 1            |

the left-most column, the constraints staisfy like this `(0x0 + 0x0) - (0x0 + (0 * (1 << 4))) = 0`.

In the next iteration, move the cursor to the next accumulate column, `accumulate_1`. 
add `add_accries_0` to previous accumulate number `0xe` then check equality with updated accumulate number `0xf` at `updated_acc` accumulate_1[1]

| -             | new_value | left_most_inv | add_carries_2 | add_carries_1 | add_carries_0 | accumulate_2 | >accumulate_1< | accumulate_0 | add_selector |
| --            | --        | --            | --            | --            | --            | --           | --           | -            | -            |
| previous_acc  |           |               |               |               |               | 0            | 0xe ←        | 0xd          | 0            | 
| updated_acc   | 0x5       | 0             | 0             | → 0           | 1 ←           | 0            | → 0xf        | 1            | 1            |

In here, the constraint satisfy that like this `(0xe + 0x1) - (0xf + (0 * (1 << 4))) = 0`